src/server/snapshot.h

```diff
@@ -40,76 +66,123 @@ class SliceSnapshot {
+  void OnDbChange(DbIndex db_index, const DbSlice::ChangeReq& req);
+
+  // Journal listener
+  void OnJournalEntry(const journal::Entry& entry);
+
+  // Close dest channel if not closed yet.
+  void CloseRecordChannel();
+
+  // Call PushFileToChannel on default buffer if needed.
+  // Flush regradless of size if force is true.
+  // Return if flushed.
+  bool FlushDefaultBuffer(bool force);
+
+  // Convert value into DbRecord.
+  DbRecord GetDbRecord(DbIndex db_index, std::string value, unsigned num_records);
+
+  // Allows getting a RdbSerializer pointer without taking care of it.
+  // Initializes a temporary serializer if the default is used on another db index.
+  // Else re-uses the default serializer.
+  struct BorrowedSerializer {
+    BorrowedSerializer(SliceSnapshot* snapshot, DbIndex, CompressionMode compression);
+    ~BorrowedSerializer();  // Flush temporary buffer or update default's stats
+
+    operator RdbSerializer*();
+
+    unsigned num_records;
+
+   private:
+    DbIndex db_index;
+    SliceSnapshot* snapshot;
+    std::optional<RdbSerializer> serializer;
+  };
+
+  friend struct BorrowedSerializer;
+
+ public:
   uint64_t snapshot_version() const {
     return snapshot_version_;
   }
 
   RdbSerializer* serializer() {
-    return rdb_serializer_.get();
+    return default_serializer_.get();
   }
 
   size_t channel_bytes() const {
-    return channel_bytes_;
+    return stats_.channel_bytes;
   }
 
   const RdbTypeFreqMap& freq_map() const {
     return type_freq_map_;
   }
 
  private:
-  void CloseRecordChannel();
-  void SerializeEntriesFb(const Cancellation* cll);
-
-  void SerializeSingleEntry(DbIndex db_index, const PrimeKey& pk, const PrimeValue& pv,
-                            RdbSerializer* serializer);
-
-  bool FlushSfile(bool force);
-  bool SaveCb(PrimeIterator it);
-
-  void OnDbChange(DbIndex db_index, const DbSlice::ChangeReq& req);
-  void OnJournalEntry(const journal::Entry& entry);
-
-  // Returns number of entries serialized.
-  // Updates the version of the bucket to snapshot version.
-  unsigned SerializePhysicalBucket(DbIndex db_index, PrimeTable::bucket_iterator it);
-  DbRecord GetDbRecord(DbIndex db_index, std::string value, unsigned num_records);
-  void PushFileToChannel(io::StringFile* sfile, DbIndex db_index, unsigned num_records,
-                         bool should_compress);
-
   DbSlice* db_slice_;
   DbTableArray db_array_;
-  RdbTypeFreqMap type_freq_map_;
 
-  std::unique_ptr<io::StringFile> sfile_;
-  std::unique_ptr<RdbSerializer> rdb_serializer_;
   RecordChannel* dest_;
+  std::atomic_bool closed_chan_{false};  // true if dest_->StartClosing was already called
 
-  boost::fibers::mutex mu_;
-  // version upper bound for entries that should be saved (not included).
-  uint64_t snapshot_version_ = 0;
-  DbIndex savecb_current_db_;  // used by SaveCb
+  DbIndex current_db_;
+  uint32_t default_buffer_records_ = 0;             // number of records in default buffer
```

Comment:
I see that there is no use in our code for DbRecord.num_records field. Maybe we can remove it and drop all the logic for the serializer number of records

Reply:
indeed